pyrdp/parser/rdp/orders/primary.py

```diff
@@ -0,0 +1,1231 @@
+        self.ctx = ctx
+
+        # Blit rectangle.
+        self.left = self.top = self.width = self.height = 0
+
+        # Source buffer offsets.
+        self.xSrc = self.ySrc = 0
+
+        self.cacheIndex = 0
+        self.cacheId = 0
+        self.colorIndex = 0
+
+    def update(self, s: BytesIO) -> 'MemBlt':
+        ctx = self.ctx
+
+        if ctx.field(1):
+            self.cacheId = Uint16LE.unpack(s)
+        if ctx.field(2):
+            self.left = read_coord(s, ctx.deltaCoords, self.left)
+        if ctx.field(3):
+            self.top = read_coord(s, ctx.deltaCoords, self.top)
+        if ctx.field(4):
+            self.width = read_coord(s, ctx.deltaCoords, self.width)
+        if ctx.field(5):
+            self.height = read_coord(s, ctx.deltaCoords, self.height)
+        if ctx.field(6):
+            self.rop = Uint8.unpack(s)
+        if ctx.field(7):
+            self.xSrc = read_coord(s, ctx.deltaCoords, self.xSrc)
+        if ctx.field(8):
+            self.ySrc = read_coord(s, ctx.deltaCoords, self.ySrc)
+        if ctx.field(9):
+            self.cacheIndex = Uint16LE.unpack(s)
+
+        self.colorIndex = self.cacheId >> 8
+        self.cacheId = self.cacheId & 0xFF
+
+        return self
+
+    def __str__(self):
+        return (f'<MemBlt ({self.xSrc},{self.ySrc}) OP={self.rop} L={self.left} T={self.top} W={self.width} H={self.height}'
+                f' cacheIndex={self.cacheIndex} cacheId={self.cacheId} colorIdx={self.colorIndex}>')
+
+
+class Mem3Blt:
+    def __init__(self, ctx: PrimaryContext):
+        self.ctx = ctx
+        self.brush = Brush()
+
+        self.cacheId = 0
+        self.left = 0
+        self.top = 0
+        self.width = 0
+        self.height = 0
+        self.rop = 0
+        self.nXSrc = 0
+        self.nYSrc = 0
+        self.bg = 0
+        self.fg = 0
+        self.cacheIndex = 0
+        self.colorIndex = 0
+        self.cacheId = 0
+
+    def update(self, s: BytesIO):
+        if self.ctx.field(1):
+            self.cacheId = Uint16LE.unpack(s)
+        if self.ctx.field(2):
+            self.left = read_coord(s, self.ctx.deltaCoords, self.left)
+        if self.ctx.field(3):
+            self.top = read_coord(s, self.ctx.deltaCoords, self.top)
+        if self.ctx.field(4):
+            self.width = read_coord(s, self.ctx.deltaCoords, self.width)
+        if self.ctx.field(5):
+            self.height = read_coord(s, self.ctx.deltaCoords, self.height)
+        if self.ctx.field(6):
+            self.rop = Uint8.unpack(s)
+        if self.ctx.field(7):
+            self.nXSrc = read_coord(s, self.ctx.deltaCoords, self.nXSrc)
+        if self.ctx.field(8):
+            self.nYSrc = read_coord(s, self.ctx.deltaCoords, self.nYSrc)
+        if self.ctx.field(9):
+            self.bg = read_rgb(s)
+        if self.ctx.field(10):
+            self.fg = read_rgb(s)
+
+        self.brush.update(s, self.ctx.fieldFlags >> 10)
+
+        if self.ctx.field(16):
```

Comment:
An enum for what? the field numbers? I don't think they are ever going to change order, and it would add one enum per message, and not really make the code more readable. It would just add like 300-400+ LoCs.

I did contemplate having some kind of loop, or defining the field handlers in an array, but it felt overengineered for no reason and even confusing. Macros would actually have been nice to have here.

Reply:
In case it's not clear what this code is doing, each field in the message is optional, and the message contains a bitfield that indicates the presence of field N with its Nth bit. Of course that bitfield is also compressed using the 2 MSB to indicate how many bytes of the bitfield are set to all 0.

This check has to be done for every field, for every message.